From 2d029a7cff942ef6d8ea97cd1e47399a6566d152 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Thu, 29 May 2025 14:07:35 +0800
Subject: [PATCH 228/254] mtk: hostapd: reset sta_info after link removing

Information like "mld_assoc_link_id", "mld_assoc_sta", and "mld_info"
need updated when AP removes the link that is some STAs' setup link.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/hostapd.c  |  1 +
 src/ap/sta_info.c | 98 +++++++++++++++++++++++++++++++++++++++++++++++
 src/ap/sta_info.h |  1 +
 3 files changed, 100 insertions(+)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index b032629cb..986aee784 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -477,6 +477,7 @@ static void hostapd_link_remove_timeout_handler(void *eloop_data,
 
 	wpa_printf(MSG_DEBUG, "MLD: Remove link_id=%u", hapd->mld_link_id);
 
+	hostapd_reset_assoc_stas(hapd);
 	hostapd_free_link_stas(hapd);
 	hostapd_disable_iface(hapd->iface);
 	mld->removed_links &= ~BIT(link_id);
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index 10338e1de..1591aea67 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -2157,3 +2157,101 @@ void ap_cleanup_sta(struct hostapd_data *hapd, const u8 *addr) {
 	if (sta)
 		ap_free_sta(hapd, sta);
 }
+
+static void hostapd_reset_assoc_sta(struct hostapd_data *hapd, struct sta_info *sta)
+{
+	struct hostapd_data *link_bss, *new_assoc_hapd = NULL;
+	struct sta_info *link_sta, *new_assoc_sta = NULL;
+	int i;
+
+	for_each_mld_link(link_bss, hapd) {
+		if (link_bss == hapd)
+			continue;
+
+		link_sta = ap_get_sta(link_bss, sta->addr);
+		if (!link_sta)
+			continue;
+
+		/* clear the flag WLAN_STA_ASSOC so that ap_get_sta() always
+		 * returns the sta_info belonging to the input hapd
+		 */
+		link_sta->flags &= ~WLAN_STA_ASSOC;
+	}
+
+	for_each_mld_link(link_bss, hapd) {
+		if (link_bss == hapd)
+			continue;
+
+		link_sta = ap_get_sta(link_bss, sta->addr);
+		if (!link_sta)
+			continue;
+
+		if (!new_assoc_hapd) {
+			new_assoc_hapd = link_bss;
+			new_assoc_sta = link_sta;
+			link_sta->eapol_sm = sta->eapol_sm;
+			os_memcpy(&link_sta->neg_ttlm, &sta->neg_ttlm,
+				  sizeof(sta->neg_ttlm));
+			wpa_reset_assoc_sm_info(new_assoc_sta->wpa_sm,
+						new_assoc_hapd->wpa_auth,
+						new_assoc_hapd->mld_link_id);
+
+			if (sta->flags & WLAN_STA_WDS) {
+				sta->flags &= ~WLAN_STA_WDS;
+				new_assoc_sta->flags |= WLAN_STA_WDS;
+			}
+
+			wpa_printf(MSG_DEBUG,
+				   "MLD: STA "MACSTR" new assoc link_id=%u",
+				   MAC2STR(sta->addr),
+				   new_assoc_hapd->mld_link_id);
+			break;
+		}
+	}
+
+	if (!new_assoc_hapd)
+		return;
+
+	for_each_mld_link(link_bss, hapd) {
+		link_sta = ap_get_sta(link_bss, sta->addr);
+		if (!link_sta)
+			continue;
+
+		link_sta->mld_assoc_link_id = new_assoc_hapd->mld_link_id;
+		link_sta->mld_assoc_sta = new_assoc_sta;
+		os_memcpy(&link_sta->mld_info, &sta->mld_info, sizeof(sta->mld_info));
+		for (i = 0; i < MAX_NUM_MLD_LINKS; i++) {
+			struct mld_link_info *li = &link_sta->mld_info.links[i];
+
+			li->resp_sta_profile = NULL;
+			li->resp_sta_profile_len = 0;
+		}
+	}
+
+	for_each_mld_link(link_bss, hapd) {
+		link_sta = ap_get_sta(link_bss, sta->addr);
+		if (!link_sta || link_sta == sta)
+			continue;
+
+		link_sta->flags |= WLAN_STA_ASSOC;
+	}
+}
+
+void hostapd_reset_assoc_stas(struct hostapd_data *hapd)
+{
+#ifdef CONFIG_IEEE80211BE
+	struct sta_info *sta, *prev;
+
+	sta = hapd->sta_list;
+	while (sta) {
+		prev = sta;
+		sta = sta->next;
+
+		if (ap_sta_is_mld(hapd, prev) &&
+		    prev->mld_assoc_link_id == hapd->mld_link_id) {
+			prev->flags &= ~WLAN_STA_ASSOC;
+			hostapd_reset_assoc_sta(hapd, prev);
+		}
+	}
+#endif /* CONFIG_IEEE80211BE */
+}
diff --git a/src/ap/sta_info.h b/src/ap/sta_info.h
index edbdbfc50..7c128bd3c 100644
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -443,5 +443,6 @@ void clear_wpa_sm_for_each_partner_link(struct hostapd_data *hapd,
 					struct sta_info *psta);
 bool ap_sta_in_list(struct hostapd_data *hapd, struct sta_info *sta);
 void ap_cleanup_sta(struct hostapd_data *hapd, const u8 *addr);
+void hostapd_reset_assoc_stas(struct hostapd_data *hapd);
 
 #endif /* STA_INFO_H */
-- 
2.45.2

